kratos/containers/constraint_data.h

```diff
@@ -0,0 +1,352 @@
+#include <unordered_set>
+#include <iostream>
+#include <assert.h>
+
+// project includes
+#include "includes/define.h"
+#include "includes/dof.h"
+#include "includes/node.h"
+#include <boost/functional/hash.hpp>
+
+namespace Kratos
+{
+
+// ConstraintEquation class start
+/*
+*   This class stores the information regarding the constraint equation. Naming convenction is defined like this.
+*   (each object of this class will store one equation in the given form)
+*
+*   slaveDOF = w_1*masterDOF_1 + w_2*masterDOF_2 + ..... + w_n*masterDOF_n
+*   
+*   each slaveDOF and each of its masterDOF have the following attributes 
+*   a. dof ID
+*   b. dof KEY
+*   c. equation ID
+*
+*   one should be able to access the constraint equation both with (dofID , dofKEY) pair and/or equationID of the slave.
+*   This equation is imposed on the linear system of equations either element wise or on the global system.
+*
+*/
+class ConstraintEquation
+{
+  private:
+    typedef Dof<double> DofType;
+    struct MasterData
+    {
+        KRATOS_CLASS_POINTER_DEFINITION(MasterData);
+        MasterData(DofType const &rMasterDof, double Weight = 0.0) : mMasterWeight(Weight), mId(rMasterDof.Id()), mKey(rMasterDof.GetVariable().Key())
+        {
+        }
+        size_t MasterKey() { return mKey; }
+        double MasterWeight() { return mMasterWeight; }
+        void UpdateWeight(double iWeight) { mMasterWeight += iWeight; }
+        size_t MasterDofId() { return mId; }
+        void SetEquationId(size_t Id) { mEquationId = Id; }
+        size_t MasterEqId() { return mEquationId; }
+        void SetMasterEqId(size_t Id) { mEquationId = Id; }
+
+      private:
+        double mMasterWeight;
+        const size_t mId;
+        const size_t mKey;
+        size_t mEquationId;
+    };
+    typedef MasterData::Pointer MasterDataPointerType;
+    struct MasterHasher
+    {
+        size_t
+        operator()(const MasterDataPointerType &rObj) const
+        {
+            std::size_t seed = 0;
+            boost::hash_combine(seed, rObj->MasterDofId());
+            boost::hash_combine(seed, rObj->MasterKey());
+            return seed;
+        }
+    };
+
+    // Custom comparator that compares the string objects by length
+    struct MasterComparator
+    {
+        bool
+        operator()(const MasterDataPointerType &rObj1, const MasterDataPointerType &rObj2) const
+        {
+            if ((rObj1->MasterDofId() == rObj2->MasterDofId()) && (rObj1->MasterKey() == rObj2->MasterKey()))
+                return true;
+            return false;
+        }
+    };
+
+    std::unordered_set<MasterDataPointerType, MasterHasher, MasterComparator> mMasterDataVector;
+
+    const size_t mId;
+    const size_t mKey;
+    size_t mEquationId;
+    double mConstant;
+    double mConstantUpdate;
+
+  public:
+    KRATOS_CLASS_POINTER_DEFINITION(ConstraintEquation);
+    typedef std::unordered_set<MasterDataPointerType, MasterHasher, MasterComparator>::iterator iterator;

```

Comment:
you mean for MasterDataPointerType ?

Reply:
Yes

Reply:
Ok. Then what happens when a pointer of MasterData is used by taking it out of the set ? For example like this 
`{
auto p_master_data = my_data_set(master) 
// do some thing with master_data
}`

here p_master_data is a unique pointer and when it is destroyed after the scope, the data is also destroyed ? This should not happen ! Please correct me if I am wrong !

Reply:
It would need to be tested but probably `auto p_master_data = my_data_set(master).get()` or `auto& r_master_data = *my_data_set(master)`.

Reply:
The both ways will work. 
First way 
`auto p_master_data = my_data_set(master).get() ` 
this would give a raw pointer right ? 

Second way
`auto& r_master_data = *my_data_set(master)`
this would mean that one should always use a it with a reference.

Reply:
Yes, from what I understand shared ptrs or unique ptrs should be used to indicate/pass ownership. So it should be correct to use a raw ptr or better a reference accessing the object to do something with it. The reason for unique ptr is that the guidelines seem to recommend this over shared ptr when there is a unique owner (in this case the set). It should avoid the overhead of needing to keep track of all the owners and simplify the internal logic.

Reply:
Besides that I didn't use unique ptr very much up to now so it is good to test if it works as you need.

Reply:
Hi guys,
            the big problem with SharedPtr is that if you start passing it around without being extra careful, you will never be able to destroy the object. So that's a sort of silent leak.

Unique ptr makes this impossible, so if you can use it i would be very much in favor of doing it.

Also, we have been making a quite big effort not to expose raw pointers around the kratos. for the reasons my hint would be to expose references (instead of raw pointers) whenever you can

Reply:
I am ok with using unique pointers here. Also it fits to the design. 
A couple of things have to be addressed here. 
1. Kratos does not have a unique_pointer yet. So one has to use std::unique_pointer
2. make_unique (like make_shared) is not possible to use when using std::unique_pointer. This is because make_unique is a available only from c++14. Though there is a way to bypass make_unique, its is ugly (making raw pointer then assigning it to unique_ptr). 

@RiccardoRossi  @pooyan-dadvand  is there a plan to include unique pointers in kratos via boost ?
I already experimented with unique pointers by using boost via shared_pointers.h. This works.